File: baselib/src/main/java/com/module/arch/base/BaseRequest.java
package com.module.arch.base;

import com.module.arch.callback.ResponseListener;

import org.jetbrains.annotations.NotNull;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.HttpException;
import retrofit2.Response;

/**
 * 请求网络封装类
 *
 * @author bd
 * @date 2021/10/09
 */
public class BaseRequest {

    private Call<Object> baseResponseCall;
    private volatile boolean disposed;

    public static BaseRequest getInstance() {
        return BaseRequest.SingletonEnum.INSTANCE.getInstance();
    }

    public void executeAsyncCallback(Call<Object> call, ResponseListener responseListener) {
        baseResponseCall = call;
        disposed = false;
        baseResponseCall.enqueue(new Callback<Object>() {
            @Override
            public void onResponse(@NotNull Call<Object> call, @NotNull Response<Object> response) {
                if (response.isSuccessful()) {
                    if (null != response.body()) {
                        if (disposed) {
                            baseResponseCall = null;
                            return;
                        }
                        responseListener.onSuccess(response.body());
                    }
                } else {
                    Throwable t = new HttpException(response);
                    responseListener.onFailure(t);
                }
                baseResponseCall = null;
            }

            @Override
            public void onFailure(@NotNull Call<Object> call, @NotNull Throwable t) {
                responseListener.onFailure(t);
                baseResponseCall = null;
            }
        });
    }

    public void cancel() {
        if (null != baseResponseCall && !baseResponseCall.isCanceled()) {
            baseResponseCall.cancel();
            baseResponseCall = null;
            disposed = true;
        }
    }

    private enum SingletonEnum {
        // 枚举对象
        INSTANCE;
        private final BaseRequest singleton;

        SingletonEnum() {
            singleton = new BaseRequest();
        }

        BaseRequest getInstance() {
            return singleton;
        }
    }
}

File: networklib/src/main/java/com/module/network/BaseRequest.kt
package com.module.network

import retrofit2.Call
import retrofit2.Callback
import retrofit2.HttpException
import retrofit2.Response

/**
 * 请求网络封装类
 *
 * @author bd
 * @date 2021/10/09
 */
class BaseRequest {

    private var baseResponseCall: Call<Any>? = null
    @Volatile
    private var disposed = false

    fun executeAsyncCallback(call: Call<Any>?, responseListener: ResponseListener) {
        baseResponseCall = call
        disposed = false
        baseResponseCall!!.enqueue(object : Callback<Any?> {
            override fun onResponse(call: Call<Any?>, response: Response<Any?>) {
                if (response.isSuccessful) {
                    if (null != response.body()) {
                        if (disposed) {
                            return
                        }
                        responseListener.onSuccess(response.body()!!)
                        baseResponseCall = null
                    }
                } else {
                    val t: Throwable = HttpException(response)
                    responseListener.onFailure(t)
                    baseResponseCall = null
                }
            }

            override fun onFailure(call: Call<Any?>, t: Throwable) {
                responseListener.onFailure(t)
                baseResponseCall = null
            }
        })
    }

    fun cancel() {
        if (null != baseResponseCall && !baseResponseCall!!.isCanceled) {
            baseResponseCall!!.cancel()
            baseResponseCall = null
            disposed = true
        }
    }

    private enum class SingletonEnum {
        // 枚举对象
        INSTANCE;

        val instance: BaseRequest = BaseRequest()
    }

    companion object {
        @JvmStatic
        val instance: BaseRequest get() = SingletonEnum.INSTANCE.instance
    }
}